### ISSUE TITLE
📊 Phase 23.3 — UtilityComputer: multi-attribute utility theory, prospect theory & preference learning

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#532]
## Overview

**UtilityComputer** computes utility values for decision alternatives using multiple frameworks: Von Neumann–Morgenstern expected utility, Kahneman–Tversky prospect theory (with loss aversion and probability weighting), and multi-attribute utility theory (MAUT) for multi-criteria decisions. Includes preference learning from historical choices via gradient descent on pairwise comparisons.

### Design Principles

| Principle | Realisation |
|-----------|-------------|
| **Behavioural realism** | Prospect theory captures loss aversion & probability distortion |
| **Multi-criteria** | MAUT weighted-sum with normalised attributes |
| **Adaptive** | Preference learning updates weights from observed choices |
| **Framework agnostic** | Single Protocol, multiple frameworks behind one API |
| **Zero-dependency** | stdlib + dataclasses; numpy optional for batch ranking |

---

### UtilityFramework (enum)

```python
class UtilityFramework(str, Enum):
    EXPECTED_UTILITY  = "expected_utility"   # VNM expected utility
    PROSPECT_THEORY   = "prospect_theory"    # Kahneman-Tversky
    MAUT              = "maut"               # Multi-attribute utility theory
    MAXIMIN           = "maximin"            # Wald criterion
    MINIMAX_REGRET    = "minimax_regret"     # Savage criterion
```

### Alternative (frozen dataclass)

```python
@dataclass(frozen=True)
class Alternative:
    id: str                                  # unique identifier
    attributes: dict[str, float]             # named attribute scores
    outcomes: tuple[float, ...]              # possible monetary/utility outcomes
    probabilities: tuple[float, ...]         # corresponding probabilities (must sum ≈ 1.0)

    def __post_init__(self) -> None:
        if len(self.outcomes) != len(self.probabilities):
            raise ValueError("outcomes and probabilities must have equal length")
        if abs(sum(self.probabilities) - 1.0) > 1e-6:
            raise ValueError("probabilities must sum to 1.0")
```

### UtilityResult (frozen dataclass)

```python
@dataclass(frozen=True)
class UtilityResult:
    alternative_id: str
    utility: float                            # computed utility value
    framework: UtilityFramework               # which framework produced this
    attribute_contributions: dict[str, float] | None  # per-attribute breakdown (MAUT only)
    prospect_value: float | None              # prospect theory value (PT only)
    certainty_equivalent: float               # CE: sure amount with same utility
```

### PreferenceModel (frozen dataclass)

```python
@dataclass(frozen=True)
class PreferenceModel:
    attribute_weights: dict[str, float]       # MAUT weights (sum to 1.0)
    loss_aversion_lambda: float = 2.25        # Tversky-Kahneman median estimate
    curvature_alpha: float = 0.88             # value function curvature
    probability_weight_gamma: float = 0.61    # Prelec weighting parameter
```

### UtilityConfig (frozen dataclass)

```python
@dataclass(frozen=True)
class UtilityConfig:
    framework: UtilityFramework = UtilityFramework.PROSPECT_THEORY
    normalize_attributes: bool = True         # min-max normalise attributes before MAUT
    learning_rate: float = 0.01               # gradient step for preference learning
    preference_history_size: int = 1000       # max pairwise comparisons to store
```

---

## Protocol

```python
@runtime_checkable
class UtilityComputer(Protocol):
    """Compute utility values for decision alternatives."""

    async def compute(
        self,
        alternative: Alternative,
        reference_point: float = 0.0,
    ) -> UtilityResult:
        """Compute utility of a single alternative.

        Args:
            alternative: The decision alternative to evaluate.
            reference_point: Reference point for prospect theory (status quo).

        Returns:
            UtilityResult with computed utility value.
        """
        ...

    async def rank(
        self,
        alternatives: Sequence[Alternative],
    ) -> Sequence[UtilityResult]:
        """Rank alternatives by utility (descending).

        Returns:
            Sequence of UtilityResult sorted by utility (highest first).
        """
        ...

    async def learn_preferences(
        self,
        chosen: Alternative,
        rejected: Alternative,
    ) -> None:
        """Update preference model from an observed pairwise choice.

        Uses gradient descent: if model ranks rejected > chosen,
        adjust weights to increase chosen utility and decrease rejected.
        """
        ...

    def get_preference_model(self) -> PreferenceModel:
        """Return current preference model snapshot."""
        ...
```

---

#### 1. Prospect Theory Value Function

```python
def _value(self, x: float, ref: float) -> float:
    """Kahneman-Tversky value function."""
    delta = x - ref
    alpha = self._prefs.curvature_alpha
    lam = self._prefs.loss_aversion_lambda
    if delta >= 0:
        return delta ** alpha
    else:
        return -lam * ((-delta) ** alpha)
```

#### 2. Prelec Probability Weighting

```python
def _weight(self, p: float) -> float:
    """Prelec (1998) probability weighting function.

    w(p) = exp(-(-ln(p))^γ)
    Overweights small probabilities, underweights large ones.
    """
    if p <= 0.0:
        return 0.0
    if p >= 1.0:
        return 1.0
    gamma = self._prefs.probability_weight_gamma
    return math.exp(-((-math.log(p)) ** gamma))
```

#### 3. Prospect Theory Utility

```python
async def _prospect_utility(self, alt: Alternative, ref: float) -> float:
    """V = Σ w(pᵢ) · v(xᵢ - ref)"""
    return sum(
        self._weight(p) * self._value(x, ref)
        for x, p in zip(alt.outcomes, alt.probabilities)
    )
```

#### 4. Expected Utility (VNM)

```python
async def _expected_utility(self, alt: Alternative) -> float:
    """EU = Σ pᵢ · u(xᵢ), where u(x) = x^α (risk aversion)."""
    alpha = self._prefs.curvature_alpha
    return sum(
        p * (x ** alpha if x >= 0 else -((-x) ** alpha))
        for x, p in zip(alt.outcomes, alt.probabilities)
    )
```

#### 5. MAUT Weighted Sum

```python
async def _maut_utility(self, alt: Alternative) -> tuple[float, dict[str, float]]:
    """U = Σ wⱼ · uⱼ(xⱼ) with optional min-max normalisation."""
    contributions: dict[str, float] = {}
    total = 0.0
    for attr, weight in self._prefs.attribute_weights.items():
        raw = alt.attributes.get(attr, 0.0)
        normalised = self._normalize(attr, raw) if self._config.normalize_attributes else raw
        contrib = weight * normalised
        contributions[attr] = contrib
        total += contrib
    return total, contributions
```

#### 6. Maximin (Wald) Criterion

```python
async def _maximin_utility(self, alt: Alternative) -> float:
    """Pessimistic: utility = min outcome."""
    return min(alt.outcomes)
```

#### 7. Minimax Regret (Savage) Criterion

```python
async def _minimax_regret(self, alt: Alternative, best_per_state: Sequence[float]) -> float:
    """Regret = max_s(best_s - outcome_s); utility = -regret."""
    regrets = [best - out for best, out in zip(best_per_state, alt.outcomes)]
    return -max(regrets)
```

#### 8. Preference Learning (Pairwise Gradient)

```python
async def learn_preferences(self, chosen: Alternative, rejected: Alternative) -> None:
    """Bradley-Terry inspired pairwise learning.

    If P(chosen > rejected) = σ(U(chosen) - U(rejected)),
    gradient on attribute weights wⱼ:
        Δwⱼ = η · (1 - σ(ΔU)) · (chosen_j - rejected_j)
    Also updates loss_aversion_lambda towards empirical ratio.
    """
    self._history.append((chosen.id, rejected.id))
    if len(self._history) > self._config.preference_history_size:
        self._history.popleft()

    u_chosen = await self.compute(chosen)
    u_rejected = await self.compute(rejected)
    delta_u = u_chosen.utility - u_rejected.utility

    # Sigmoid
    prob = 1.0 / (1.0 + math.exp(-delta_u))
    error = 1.0 - prob  # should be ≈0 if model agrees with choice
    lr = self._config.learning_rate

    # Update attribute weights
    new_weights = {}
    for attr, w in self._prefs.attribute_weights.items():
        grad = error * (chosen.attributes.get(attr, 0.0) - rejected.attributes.get(attr, 0.0))
        new_weights[attr] = max(0.0, w + lr * grad)

    # Re-normalise weights to sum to 1.0
    total = sum(new_weights.values()) or 1.0
    new_weights = {k: v / total for k, v in new_weights.items()}

    self._prefs = PreferenceModel(
        attribute_weights=new_weights,
        loss_aversion_lambda=self._prefs.loss_aversion_lambda,
        curvature_alpha=self._prefs.curvature_alpha,
        probability_weight_gamma=self._prefs.probability_weight_gamma,
    )
    self._pref_updates_counter.inc()
```

#### 9. Certainty Equivalent

```python
def _certainty_equivalent(self, utility: float) -> float:
    """Inverse of value function: CE such that v(CE) = utility."""
    alpha = self._prefs.curvature_alpha
    if utility >= 0:
        return utility ** (1.0 / alpha)
    else:
        return -((-utility / self._prefs.loss_aversion_lambda) ** (1.0 / alpha))
```

### Framework Dispatch

```python
async def compute(self, alternative: Alternative, reference_point: float = 0.0) -> UtilityResult:
    match self._config.framework:
        case UtilityFramework.EXPECTED_UTILITY:
            u = await self._expected_utility(alternative)
            return UtilityResult(alternative.id, u, self._config.framework,
                                 None, None, self._certainty_equivalent(u))
        case UtilityFramework.PROSPECT_THEORY:
            pv = await self._prospect_utility(alternative, reference_point)
            return UtilityResult(alternative.id, pv, self._config.framework,
                                 None, pv, self._certainty_equivalent(pv))
        case UtilityFramework.MAUT:
            u, contribs = await self._maut_utility(alternative)
            return UtilityResult(alternative.id, u, self._config.framework,
                                 contribs, None, u)
        case UtilityFramework.MAXIMIN:
            u = await self._maximin_utility(alternative)
            return UtilityResult(alternative.id, u, self._config.framework,
                                 None, None, u)
        case _:
            raise ValueError(f"unsupported framework: {self._config.framework}")
    self._computations_counter.inc()
    self._framework_counter.labels(framework=self._config.framework.value).inc()
```

### Ranking

```python
async def rank(self, alternatives: Sequence[Alternative]) -> Sequence[UtilityResult]:
    """Compute utility for all alternatives, sort descending."""
    with self._ranking_duration.time():
        results = await asyncio.gather(*(self.compute(a) for a in alternatives))
        return sorted(results, key=lambda r: r.utility, reverse=True)
```

---

## NullUtilityComputer

```python
class NullUtilityComputer:
    """No-op implementation for DI / testing."""

    async def compute(self, alternative: Alternative, reference_point: float = 0.0) -> UtilityResult:
        return UtilityResult(alternative.id, 0.0, UtilityFramework.EXPECTED_UTILITY, None, None, 0.0)

    async def rank(self, alternatives: Sequence[Alternative]) -> Sequence[UtilityResult]:
        return [await self.compute(a) for a in alternatives]

    async def learn_preferences(self, chosen: Alternative, rejected: Alternative) -> None:
        pass

    def get_preference_model(self) -> PreferenceModel:
        return PreferenceModel(attribute_weights={})
```

---

## Factory

```python
def make_utility_computer(
    config: UtilityConfig | None = None,
    initial_weights: dict[str, float] | None = None,
    *,
    null: bool = False,
) -> UtilityComputer:
    if null:
        return NullUtilityComputer()
    cfg = config or UtilityConfig()
    weights = initial_weights or {}
    prefs = PreferenceModel(attribute_weights=weights)
    return AdaptiveUtilityComputer(cfg, prefs)
```

---

## Prometheus Metrics

| Metric | Type | Description |
|--------|------|-------------|
| `asi_utility_computations_total` | Counter | Total utility computations |
| `asi_utility_framework_usage` | Counter (label: framework) | Computations per framework |
| `asi_utility_preference_updates_total` | Counter | Pairwise preference learning updates |
| `asi_utility_loss_aversion_lambda` | Gauge | Current loss aversion parameter |
| `asi_utility_ranking_duration_seconds` | Histogram | Time to rank a set of alternatives |

### PromQL Examples

```promql
# Computation rate by framework
rate(asi_utility_framework_usage[5m])

# Loss aversion drift over time
asi_utility_loss_aversion_lambda

# Average ranking latency
histogram_quantile(0.95, rate(asi_utility_ranking_duration_seconds_bucket[5m]))
```

### Grafana Alerts

```yaml
- alert: UtilityComputationSpikeAlert
  expr: rate(asi_utility_computations_total[5m]) > 500
  for: 2m
  annotations:
    summary: "Utility computation rate spike (>500/s for 2m)"

- alert: LossAversionDriftAlert
  expr: abs(asi_utility_loss_aversion_lambda - 2.25) > 0.5
  for: 10m
  annotations:
    summary: "Loss aversion λ drifted >0.5 from Kahneman-Tversky baseline 2.25"

- alert: RankingLatencyAlert
  expr: histogram_quantile(0.95, rate(asi_utility_ranking_duration_seconds_bucket[5m])) > 0.5
  for: 5m
  annotations:
    summary: "P95 ranking latency > 500ms"
```

---

## Integration Points

| Component | Direction | Contract |
|-----------|-----------|----------|
| **DecisionFramework 23.1** | ← | Receives framework type and context for compute() |
| **ProbabilityEstimator 23.2** | ← | Provides outcome probabilities for Alternative construction |
| **StrategySelector 23.4** | → | Ranked UtilityResults feed strategy selection |
| **DecisionOrchestrator 23.5** | ↔ | Orchestrates compute→rank pipeline, triggers learn_preferences |
| **CreativeOrchestrator 22.5** | ← | Novel alternatives from divergent generation |
| **AffectiveOrchestrator 21.5** | ← | Emotional valence influences reference_point |
| **ReasoningOrchestrator 20.5** | ← | Logical constraints filter alternatives before ranking |

---

## mypy Strict Compliance

| Pattern | Technique |
|---------|-----------|
| `dict[str, float]` return | Always explicit, never `Any` |
| `tuple[float, ...]` | Homogeneous tuple for outcomes/probabilities |
| `float \| None` | Used for optional prospect_value / attribute_contributions |
| `match/case` on enum | Exhaustive with `case _: raise ValueError` |
| `@runtime_checkable` Protocol | Structural typing for DI |

---

## Test Targets (12)

| # | Test | Assertion |
|---|------|-----------|
| 1 | `test_prospect_value_loss_aversion` | Losses weighted λ× more than equivalent gains |
| 2 | `test_prelec_small_probability_overweight` | w(0.05) > 0.05 |
| 3 | `test_prelec_large_probability_underweight` | w(0.95) < 0.95 |
| 4 | `test_expected_utility_risk_neutral` | α=1.0 → EU equals expected value |
| 5 | `test_maut_weights_sum` | Attribute contributions sum to total utility |
| 6 | `test_maut_normalisation` | Normalised attributes in [0, 1] |
| 7 | `test_maximin_selects_worst_case` | Utility equals minimum outcome |
| 8 | `test_rank_ordering` | Alternatives returned in descending utility order |
| 9 | `test_learn_preferences_weight_shift` | After learning, chosen alternative weight increases |
| 10 | `test_certainty_equivalent_roundtrip` | v(CE(u)) ≈ u within tolerance |
| 11 | `test_null_returns_zero` | NullUtilityComputer returns utility=0.0 |
| 12 | `test_factory_null_flag` | `make_utility_computer(null=True)` returns NullUtilityComputer |

### Test Skeletons

```python
@pytest.mark.asyncio
async def test_prospect_value_loss_aversion():
    """Verify losses are weighted λ× more than equivalent gains."""
    config = UtilityConfig(framework=UtilityFramework.PROSPECT_THEORY)
    computer = make_utility_computer(config)
    prefs = computer.get_preference_model()

    gain_alt = Alternative("gain", {}, (100.0,), (1.0,))
    loss_alt = Alternative("loss", {}, (-100.0,), (1.0,))

    gain_result = await computer.compute(gain_alt, reference_point=0.0)
    loss_result = await computer.compute(loss_alt, reference_point=0.0)

    # Loss should be λ× the magnitude of gain
    assert abs(loss_result.utility) > abs(gain_result.utility)
    ratio = abs(loss_result.utility) / abs(gain_result.utility)
    assert abs(ratio - prefs.loss_aversion_lambda) < 0.01


@pytest.mark.asyncio
async def test_maut_attribute_weights():
    """Verify MAUT produces correct weighted-sum utility."""
    weights = {"speed": 0.6, "cost": 0.4}
    config = UtilityConfig(framework=UtilityFramework.MAUT, normalize_attributes=False)
    computer = make_utility_computer(config, initial_weights=weights)

    alt = Alternative("A", {"speed": 0.8, "cost": 0.7}, (1.0,), (1.0,))
    result = await computer.compute(alt)

    expected = 0.6 * 0.8 + 0.4 * 0.7  # 0.76
    assert abs(result.utility - expected) < 1e-6
    assert result.attribute_contributions is not None
    assert abs(result.attribute_contributions["speed"] - 0.48) < 1e-6
    assert abs(result.attribute_contributions["cost"] - 0.28) < 1e-6
```

---

## Implementation Order (14 steps)

| Step | Task | File |
|------|------|------|
| 1 | `UtilityFramework` enum | `enums.py` |
| 2 | `Alternative` frozen dataclass | `models.py` |
| 3 | `UtilityResult` frozen dataclass | `models.py` |
| 4 | `PreferenceModel` frozen dataclass | `models.py` |
| 5 | `UtilityConfig` frozen dataclass | `config.py` |
| 6 | `UtilityComputer` Protocol | `protocols.py` |
| 7 | `_value()` prospect theory value function | `adaptive.py` |
| 8 | `_weight()` Prelec probability weighting | `adaptive.py` |
| 9 | `_prospect_utility()` + `_expected_utility()` | `adaptive.py` |
| 10 | `_maut_utility()` + normalisation | `adaptive.py` |
| 11 | `_maximin_utility()` + `_minimax_regret()` | `adaptive.py` |
| 12 | `learn_preferences()` gradient descent | `adaptive.py` |
| 13 | `NullUtilityComputer` + `make_utility_computer()` | `factory.py` |
| 14 | 12 tests | `test_utility_computer.py` |

---

## Phase 23 Sub-Phase Tracker

| Sub-Phase | Component | Issue | Status |
|-----------|-----------|-------|--------|
| 23.1 | DecisionFramework | TBD | 🔲 |
| 23.2 | ProbabilityEstimator | TBD | 🔲 |
| **23.3** | **UtilityComputer** | **this** | **🟢 SPEC** |
| 23.4 | StrategySelector | TBD | 🔲 |
| 23.5 | DecisionOrchestrator | TBD | 🔲 |
